src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -1701,16 +1707,48 @@ private IEnumerable<PSTypeName> InferTypesFrom(MemberExpressionAst memberExpress
 
         private static void AddTypesFromMethodCacheEntry(
             DotNetAdapter.MethodCacheEntry methodCacheEntry,
+            IList<ITypeName> genericTypeArguments,
             List<PSTypeName> result,
             bool isInvokeMemberExpressionAst)
         {
             if (isInvokeMemberExpressionAst)
             {
+                Type[] resolvedTypeArguments = null;
+                if (genericTypeArguments is not null)
+                {
+                    resolvedTypeArguments = new Type[genericTypeArguments.Count];
+                    for (int i = 0; i < genericTypeArguments.Count; i++)
+                    {
+                        Type resolvedType = genericTypeArguments[i].GetReflectionType();
+                        if (resolvedType is null)
+                        {
+                            return;
+                        }
+
+                        resolvedTypeArguments[i] = resolvedType;
+                    }
+                }
+
                 foreach (var method in methodCacheEntry.methodInformationStructures)
                 {
-                    if (method.method is MethodInfo methodInfo && !methodInfo.ReturnType.ContainsGenericParameters)
+                    if (method.method is MethodInfo methodInfo)
                     {
-                        result.Add(new PSTypeName(methodInfo.ReturnType));
+                        if (!methodInfo.ReturnType.ContainsGenericParameters)
+                        {
+                            result.Add(new PSTypeName(methodInfo.ReturnType));
```

Comment:
A couple things here:
1. It looks to me will add `Void` type as well, but maybe we shouldn't add `void` type?
2. This will add multiple entries for the same return type for some overload methods, for example, `$a = [ConcurrentDictionary]::new();$a.TryGetValue(xxx).<tab>`. Maybe skip the same return types so as to reduce the work later to filter those duplicates?

Reply:
1: Seems reasonable to me.
2: Agreed but the result list gets passed through several methods, maybe it would be better to just use a hashset instead of explicitly handling duplicates in every method? I've not looked through the other code to know how feasible that is but `PSTypeName` would probably need to get an `Equals` method added at a minimum.

Reply:
> result list gets passed through several methods, maybe it would be better to just use a hashset instead of explicitly handling duplicates in every method?

It sounds reasonable to use `Hashset` for `results`, but downsides are
1. that would require changes in lots of places I believe.
2. it doesn't guarantee order, so it may be possible tab completion on the same script return results in different order from different runs, and I guess that's not desirable.

Are there other places where duplicate entries are likely to be added to the result list like here?

Reply:
I haven't really looked at the code to know for sure if it's a problem elsewhere. I guess we can start by fixing it here and wait for a different PR to change it elsewhere if needed.